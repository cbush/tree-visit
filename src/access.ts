import { IndexPath } from './indexPath'
import { BaseOptions } from './options'

/**
 * Returns a node by its `IndexPath`.
 *
 * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
 */
export function access<T>(
  node: T,
  indexPath: IndexPath,
  options: BaseOptions<T>
): T {
  if (indexPath.length === 0) return node

  const children = options.getChildren(node, indexPath)

  return access(children[indexPath[0]], indexPath.slice(1), options)
}

/**
 * Returns an array of each node in an `IndexPath`.
 *
 * The first node is implicitly included in the `IndexPath` (i.e. no need to pass a `0` first in every `IndexPath`).
 */
export function accessPath<T>(
  node: T,
  indexPath: IndexPath,
  options: BaseOptions<T>
): T[] {
  if (indexPath.length === 0) return [node]

  const children = options.getChildren(node, indexPath)

  const result = accessPath(children[indexPath[0]], indexPath.slice(1), options)

  result.unshift(node)

  return result
}
