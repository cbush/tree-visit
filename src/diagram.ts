import { IndexPath } from './indexPath'
import { BaseOptions } from './options'
import { boxDiagram } from './diagram/boxDiagram'
import { directoryDiagram } from './diagram/directoryDiagram'

export type DiagramType = 'directory' | 'box'

export type DiagramOptions<T> = BaseOptions<T> & {
  getLabel: (node: T, indexPath: IndexPath) => string
  type?: DiagramType
}

/**
 * Generate a diagram of the tree, as a string.
 */
export function diagram<T>(node: T, options: DiagramOptions<T>): string {
  if (options.type === 'box') {
    return boxDiagram(node, options)
  }

  return directoryDiagram(node, options)
}
