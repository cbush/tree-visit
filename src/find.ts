import { IndexPath } from './indexPath'
import { BaseOptions } from './options'
import { visit, STOP } from './visit'

type FindOptions<T> = BaseOptions<T> & {
  predicate: (node: T, indexPath: IndexPath) => boolean
}

export function find<T>(node: T, options: FindOptions<T>): T | undefined {
  let found: T | undefined

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        found = child
        return STOP
      }
    },
    getChildren: options.getChildren,
  })

  return found
}

export function findAll<T>(node: T, options: FindOptions<T>): T[] {
  let found: T[] = []

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        found.push(child)
      }
    },
    getChildren: options.getChildren,
  })

  return found
}

export function findIndexPath<T>(
  node: T,
  options: FindOptions<T>
): IndexPath | undefined {
  let found: IndexPath | undefined

  visit(node, {
    onEnter: (child, indexPath) => {
      if (options.predicate(child, indexPath)) {
        // Copy the indexPath, since indexPath may be mutated
        found = [...indexPath]
        return STOP
      }
    },
    getChildren: options.getChildren,
  })

  return found
}
