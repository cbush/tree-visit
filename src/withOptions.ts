import { BaseOptions } from './options'
import { visit, VisitOptions } from './visit'
import { IndexPath } from './indexPath'
import { access, accessPath } from './access'
import { find, findAll, findIndexPath, FindOptions } from './find'

export type WithOptions<T> = {
  access(node: T, indexPath: IndexPath): T
  accessPath(node: T, indexPath: IndexPath): T[]
  find(node: T, predicate: FindOptions<T>['predicate']): T | undefined
  find(
    node: T,
    options: Omit<FindOptions<T>, keyof BaseOptions<T>>
  ): T | undefined
  findAll(node: T, predicate: FindOptions<T>['predicate']): T[]
  findAll(node: T, options: Omit<FindOptions<T>, keyof BaseOptions<T>>): T[]
  findIndexPath(
    node: T,
    predicate: FindOptions<T>['predicate']
  ): IndexPath | undefined
  findIndexPath(
    node: T,
    options: Omit<FindOptions<T>, keyof BaseOptions<T>>
  ): IndexPath | undefined
  /**
   * Visit each node using preorder DFS traversal.
   */
  visit(node: T, onEnter: NonNullable<VisitOptions<T>>['onEnter']): void
  /**
   * Visit each node using DFS traversal.
   */
  visit(node: T, options: Omit<VisitOptions<T>, keyof BaseOptions<T>>): void
}

/**
 * Return every tree utility function with options partially applied.
 *
 * @param baseOptions
 */
export function withOptions<T>(baseOptions: BaseOptions<T>): WithOptions<T> {
  return {
    access: (node, indexPath) => access(node, indexPath, baseOptions),
    accessPath: (node, indexPath) => accessPath(node, indexPath, baseOptions),
    find: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | Omit<FindOptions<T>, keyof BaseOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? find(node, { ...baseOptions, predicate: predicateOrOptions })
        : find(node, { ...baseOptions, ...predicateOrOptions }),
    findAll: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | Omit<FindOptions<T>, keyof BaseOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findAll(node, { ...baseOptions, predicate: predicateOrOptions })
        : findAll(node, { ...baseOptions, ...predicateOrOptions }),
    findIndexPath: (
      node: T,
      predicateOrOptions:
        | FindOptions<T>['predicate']
        | Omit<FindOptions<T>, keyof BaseOptions<T>>
    ) =>
      typeof predicateOrOptions === 'function'
        ? findIndexPath(node, { ...baseOptions, predicate: predicateOrOptions })
        : findIndexPath(node, { ...baseOptions, ...predicateOrOptions }),
    visit: (
      node: T,
      onEnterOrOptions:
        | NonNullable<VisitOptions<T>>['onEnter']
        | Omit<VisitOptions<T>, keyof BaseOptions<T>>
    ) =>
      typeof onEnterOrOptions === 'function'
        ? visit(node, { ...baseOptions, onEnter: onEnterOrOptions })
        : visit(node, { ...baseOptions, ...onEnterOrOptions }),
  }
}
