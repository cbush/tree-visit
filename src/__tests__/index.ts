import { visit, IndexPath } from '../index'
import { access, accessPath } from '../access'
import { find, findIndexPath, findAll } from '../find'
import { withOptions } from '../withOptions'
import { diagram, DiagramOptions } from '../diagram'

type Node = {
  name: string
  children?: Node[]
  indexPath: IndexPath
}

function getChildren(node: Node): Node[] {
  return node.children ?? []
}

const example: Node = {
  name: 'a',
  children: [
    {
      name: 'b',
      indexPath: [0],
      children: [
        { name: 'b1', indexPath: [0, 0] },
        { name: 'b2', indexPath: [0, 1] },
      ],
    },
    {
      name: 'c',
      indexPath: [1],
      children: [
        { name: 'c1', indexPath: [1, 0] },
        { name: 'c2', indexPath: [1, 1] },
      ],
    },
  ],
  indexPath: [],
}

it('traverses normally', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []

  visit(example, {
    onEnter: (child, indexPath) => {
      expect(indexPath).toEqual(child.indexPath)
      enterNames.push(child.name)
    },
    onLeave: (child, indexPath) => {
      expect(indexPath).toEqual(child.indexPath)
      leaveNames.push(child.name)
    },
    getChildren,
  })

  expect(enterNames).toEqual(['a', 'b', 'b1', 'b2', 'c', 'c1', 'c2'])
  expect(leaveNames).toEqual(['b1', 'b2', 'b', 'c1', 'c2', 'c', 'a'])
})

it('skips in onEnter', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []

  visit(example, {
    onEnter: (child) => {
      enterNames.push(child.name)
      if (child.name === 'b') return 'skip'
    },
    onLeave: (child) => {
      leaveNames.push(child.name)
    },
    getChildren,
  })

  expect(enterNames).toEqual(['a', 'b', 'c', 'c1', 'c2'])
  expect(leaveNames).toEqual(['c1', 'c2', 'c', 'a'])
})

it('stops in onEnter', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []

  visit(example, {
    onEnter: (child) => {
      enterNames.push(child.name)
      if (child.name === 'b') return 'stop'
    },
    onLeave: (child) => {
      leaveNames.push(child.name)
    },
    getChildren,
  })

  expect(enterNames).toEqual(['a', 'b'])
  expect(leaveNames).toEqual([])
})

it('stops in onLeave', () => {
  let enterNames: string[] = []
  let leaveNames: string[] = []

  visit(example, {
    onEnter: (child) => {
      enterNames.push(child.name)
    },
    onLeave: (child) => {
      leaveNames.push(child.name)
      if (child.name === 'b') return 'stop'
    },
    getChildren,
  })

  expect(enterNames).toEqual(['a', 'b', 'b1', 'b2'])
  expect(leaveNames).toEqual(['b1', 'b2', 'b'])
})

describe('access', () => {
  it('accesses node at index path', () => {
    visit(example, {
      onEnter: (child, indexPath) => {
        const found = access(example, indexPath, { getChildren })
        expect(found.name).toEqual(child.name)
      },
      getChildren,
    })
  })

  it('accesses node path', () => {
    expect(
      accessPath(example, [0, 1], { getChildren }).map((node) => node.name)
    ).toEqual(['a', 'b', 'b2'])
  })
})

describe('find', () => {
  it('finds a node', () => {
    const node = find(example, {
      getChildren,
      predicate: (node) => node.name === 'b1',
    })

    expect(node?.name).toEqual('b1')
  })

  it('finds all nodes', () => {
    const nodes = findAll(example, {
      getChildren,
      predicate: (node) => node.name.startsWith('b'),
    })

    expect(nodes.map((node) => node.name)).toEqual(['b', 'b1', 'b2'])
  })

  it('finds a node index path', () => {
    const indexPath = findIndexPath(example, {
      getChildren,
      predicate: (node, indexPath) => indexPath.join() === [0, 1].join(),
    })

    expect(indexPath).toEqual([0, 1])
  })
})

describe('diagram', () => {
  const getLabel = (node: Node): string => node.name
  const getMultilineLabel = (node: Node): string =>
    `name: ${node.name}\npath: ${node.indexPath.join('.')}`

  it('generates folder diagram', () => {
    expect(diagram(example, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('generates folder diagram with single child case', () => {
    const singleChild: Node = {
      name: 'a',
      children: [
        {
          name: 'b',
          indexPath: [0],
          children: [{ name: 'b1', indexPath: [0, 0] }],
        },
      ],
      indexPath: [],
    }

    expect(diagram(singleChild, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('generates folder diagram with hidden root', () => {
    const singleChild: Node = {
      name: '',
      children: [
        {
          name: 'b',
          indexPath: [0],
          children: [{ name: 'b1', indexPath: [0, 0] }],
        },
      ],
      indexPath: [],
    }

    expect(diagram(singleChild, { getChildren, getLabel })).toMatchSnapshot()
  })

  it('generates folder diagram with multiline label', () => {
    expect(
      diagram(example, { getChildren, getLabel: getMultilineLabel })
    ).toMatchSnapshot()
  })

  it('generates folder diagram with multiline label and single child case', () => {
    const singleChild: Node = {
      name: 'a',
      children: [
        {
          name: 'b',
          indexPath: [0],
          children: [{ name: 'b1', indexPath: [0, 0] }],
        },
      ],
      indexPath: [],
    }

    expect(
      diagram(singleChild, { getChildren, getLabel: getMultilineLabel })
    ).toMatchSnapshot()
  })
})

describe('withOptions', () => {
  it('binds options', () => {
    const { find, access, visit } = withOptions({ getChildren })

    let enterNames: string[] = []

    visit(example, {
      onEnter: (child) => {
        enterNames.push(child.name)
      },
    })

    expect(enterNames).toEqual(['a', 'b', 'b1', 'b2', 'c', 'c1', 'c2'])

    expect(
      find(example, {
        predicate: (node) => node.name === 'b1',
      })?.name
    ).toEqual('b1')

    expect(access(example, [0, 0]).name).toEqual('b1')
  })

  it('supports overloaded calls', () => {
    const { find, visit, diagram } = withOptions({ getChildren })

    let enterNames: string[] = []

    visit(example, (child) => {
      enterNames.push(child.name)
    })

    expect(enterNames).toEqual(['a', 'b', 'b1', 'b2', 'c', 'c1', 'c2'])

    expect(find(example, (node) => node.name === 'b1')?.name).toEqual('b1')

    expect(diagram(example, (node) => node.name)).toMatchSnapshot()
  })
})
