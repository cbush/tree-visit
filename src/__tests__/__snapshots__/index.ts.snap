// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`diagram generates folder diagram 1`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;

exports[`diagram generates folder diagram with hidden root 1`] = `"/ b / b1"`;

exports[`diagram generates folder diagram with multiline label 1`] = `
"name: a
path: 
├── name: b
│   path: 0
│   ├── name: b1
│   │   path: 0.0
│   └── name: b2
│       path: 0.1
└── name: c
    path: 1
    ├── name: c1
    │   path: 1.0
    └── name: c2
        path: 1.1"
`;

exports[`diagram generates folder diagram with multiline label and single child case 1`] = `
"name: a
path: 
└── name: b
    path: 0
    └── name: b1
        path: 0.0"
`;

exports[`diagram generates folder diagram with single child case 1`] = `"a / b / b1"`;

exports[`withOptions supports overloaded calls 1`] = `
"a
├── b
│   ├── b1
│   └── b2
└── c
    ├── c1
    └── c2"
`;
